### ISSUE TITLE
sidekiq web interface show Forbidden when I try delete or post operations.

### ISSUE THREAD

**Issue by jjyr** [mperham/sidekiq#1289]
Under development environment, everything is well.
But when I deployed to server, and try some operations like 'delete' or 'retry', the web interface just response "Forbidden".

**Comment by JamesChevalier** [comment]
I'm having the same issue after moving server - so it's definitely a web configuration issue, but I can't figure out what's wrong. The server move included an upgrade to apache (from 2.2.22 to 2.4.7), so there's probably something there that's causing the problem.

Is the issue obvious to anyone else? I'm having a tough time working it out...
Thanks for any help you can manage!

Old `mods-enabled`:

    actions.conf  alias.load       authz_default.load    authz_user.load  cgid.conf     deflate.load  env.load   negotiation.conf  passenger.load rewrite.load   ssl.conf     status.load
    actions.load  auth_basic.load  authz_groupfile.load  autoindex.conf   cgid.load     dir.conf    mime.conf  negotiation.load  reqtimeout.conf  setenvif.conf  ssl.load
    alias.conf    authn_file.load  authz_host.load       autoindex.load   deflate.conf  dir.load    mime.load  passenger.conf    reqtimeout.load  setenvif.load  status.conf

New `modes-enabled`:

    access_compat.load  auth_basic.load  authz_core.load  autoindex.conf  deflate.load  env.load     mime.load       negotiation.conf  passenger.load  setenvif.load
    alias.conf          authn_core.load  authz_host.load  autoindex.load  dir.conf      filter.load  mpm_event.conf  negotiation.load  rewrite.load    status.conf
    alias.load          authn_file.load  authz_user.load  deflate.conf    dir.load      mime.conf    mpm_event.load  passenger.conf    setenvif.conf   status.load


Old `Directory`:

    <Directory /home/user/production/current/public>
        AllowOverride all
        Options -MultiViews
        RailsEnv production
    </Directory>

New `Directory`:

    <Directory /home/user/production/current/public>
      Require all granted
      AllowOverride all
      Options -MultiViews
    </Directory>

**Comment by mperham** [comment]
This is the new CSRF protection in action.  You need to have a session store enabled now.

https://github.com/mperham/sidekiq/blob/master/Changes.md#342

**Comment by JamesChevalier** [comment]
Oh yeah? Even if I'm doing it in Rails routes?

    authenticate :user, lambda { |user| user.role == 'admin' } do
      mount Sidekiq::Web, at: '/sidekiq'
    end

**Comment by mperham** [comment]
I can't say for sure because every Rails app is different but if you just upgraded to 3.4.2, it's almost guaranteed that it's the new CSRF protection.

**Comment by typeoneerror** [comment]
@JamesChevalier let me know if you figure out how to get this working with Rails. I'm getting "Forbidden" white screens as well.

**Comment by typeoneerror** [comment]
I'm seeing the authenticity_token in the sidekiq ui forms at http://localhost:3000/sidekiq. Example:

`<input type='hidden' name='authenticity_token' value='67cd8bcaa60b3ffb3d8c3dd3a8c941f0'/>`

Doesn't look like a typical Rails token format though so I'm not sure what I've done wrong. Here's what they look like in a rails form at localhost:3000 (same app running on puma web server).

`<input type="hidden" name="authenticity_token" value="m23YrEfFCcKXuuFY+TmzZK+48SUTLT4Hq7kF9/77n8123snGAKU+eE+ALEtN8/pmaoxC4mfhp3l/SbpI4n/+o1A==" />`

Any idea what could be wrong?

**Comment by JamesChevalier** [comment]
My issue was due to the `csrf_tag` not being present in the [sidekiq-failures](https://github.com/mhfs/sidekiq-failures) form. We sorted that out over [there](https://github.com/mhfs/sidekiq-failures/issues/98).

I'd suggest taking a look through your sidekiq plugins - if you're using any.

**Comment by typeoneerror** [comment]
@JamesChevalier hmm, yeah, I saw that. Not using any sidekiq plugins, just the "out of the box" mount route.

**Comment by typeoneerror** [comment]
Getting somewhere. I switched to:

`Rails.application.config.session_store :cookie_store, key: '_doki'`

from:

```
Rails.application.config.session_store :cookie_store, key: '_doki', domain: {
  production: '.doki.io',
  development: '.doki.dev'
}.fetch(Rails.env.to_sym, :all)
```

and now it works. has to be something to do with the domain property.

**Comment by tanelsuurhans** [comment]
It would seem that this only works when you are using cookie_store as your session backend. My application is using Redis for sessions and deleting/retying jobs in Sidekiq flat out will not work. If I switch to cookie store, then it starts working (without specifying a domain key). I am assuming the problem here is that SidekiqWeb does not "know" how to use the Redis session backend and this breaks the CSFR protection @mperham ? Any ideas how to make it work?

**Comment by rikas** [comment]
I'm still getting errors. I have cookies session store and I'm using Sidekiq 3.5.0.

I have no idea about what's going on. It only happens in production (NGINX) and I've been changing my configuration according to some tips I came across here, but no luck. :cry:

**Comment by mperham** [comment]
Someone posted this about nginx and Forbidden:

https://github.com/mperham/sidekiq/issues/2560

**Comment by rikas** [comment]
@mperham yes, I know... I added that to my configuration but I'm still having the issue.

**Comment by tanelsuurhans** [comment]
Yeah that does not seem to change anything.

**Comment by dreyks** [comment]
Having the same issue, also using Redis for sessions. But it only happens to me on production servers and not on staging, which also uses Redis sessions.

I'll try to figure out what's the setup difference

**Comment by dreyks** [comment]
Ok. Turns out that the difference was that production has multiple app instances.

Sinatra, if given no secret for cookie session storage, generates a random one as the server starts. So I end up with different secrets on different app instances and cookies signed on one instance are not recognized on other one, which leads to new session with new csrf token, which does not match authenticity token anymore

Sidekiq::Web should probably set secret explicitly, though it should probably be configurable via ENV or sidekiq config

@mperham what do you think?

**Comment by mperham** [comment]
@dreyks Sidekiq is not in the business of creating/managing sessions.  That's why this issue has been so painful: I can't control it.  The application must provide a valid session.

**Comment by tanelsuurhans** [comment]
Its clear the functionality is broken in its current state. On one hand I agree with you that the app you mount Sidekiq::Web to should provide the session. On the other hand the default is not obviously working in all scenarios and gaining access/providing the needed details to the Sinatra app seems to be either tricky or not doable at all. We need some sort of a solution though.

**Comment by dreyks** [comment]
Well I might not completely understand, but the main app is Rails, and Sidekiq::Web is a sinatra one and it has `enable :sessions` so how can someone share sessions between apps?

**Comment by dreyks** [comment]
@tanelsuurhans yeah, this seems not to work out of the box for certain situations

**Comment by mperham** [comment]
Great question.  I wish Sinatra and Rails would work that out and document it.  Sidekiq::Web declares that it needs a session.  Beyond that, it's up to you to configure the web frameworks correctly to make it work.

**Comment by dreyks** [comment]
Well... Sinatra has a mechanism to set the secret `set :session_secret, 'super secret'`. Sidekiq could have a way to set a secret from Rails in its configure section and then use it in Sidekiq::Web

**Comment by dreyks** [comment]
@mperham as you rightly said it's up to the app to provide a valid session, but I think it's also up to Sidekiq to provide means of doing this.
How about something like this
```ruby
# lib/sidekiq.rb
def self.sinatra.session=(session)
  @sinatra_session = session
end

def self.sinatra_session
  @sinatra_session ||= -> { enable :sessions }
end

# lib/sidekiq/web.rb
# ...skipped ...
class Web < Sinatra::Base
    include Sidekiq::Paginator

    instance_exec(&Sidekiq.sinatra_session)
    use Rack::Protection, :use => :authenticity_token unless ENV['RACK_ENV'] == 'test'

# ...skipped...
end
```

Then if one wants to use say Redis for storing sessions, it can be done with something like

```ruby
# config/initializers/sidekiq.rb
Sidekiq.configure_client do |config|
  config.sinatra_session = lambda do
    require 'rack/session/redis'
    use Rack::Session::Redis
  end
end
```

This is only a sketch, if you're interested I can pull it together into a PR

**Comment by mperham** [comment]
@dreyks That's a silly amount of preamble when the user just needs to do this:

Sidekiq::Web.use Rack::Session::Redis

**Comment by dreyks** [comment]
oh... really. thanks =)

**Comment by juanibiapina** [comment]
We're having the same issue. Sidekiq Web is mounted as a Rails route, which means it will get sessions for free: (https://robots.thoughtbot.com/how-to-share-a-session-between-sinatra-and-rails)

We use a cookie store like this:

```
Rails.application.config.session_store :cookie_store, key: "_csi_session"
```

Still is does not work out of the box in production. We're using a galeb router.

I'll leave this here until someone can figure out what the real problem is.

**Comment by sainejob** [comment]
FWIW- I was encountering this issue with Rails 4.0.13, Sinatra 1.4.3, after upgrading to Sidekiq 3.5.3.

The 'authenticity_token' in the Sidekiq Web UI was not being populated.

I upgraded to Sinatra 1.4.6 (which upgraded my Rack::Protection) and the token was populated and everything started working again.

I use the following setup: 

```
require 'sidekiq/pro/web'
Sidekiq::Web.use(Rack::Auth::Basic) do |user, password|
   [user, password] == [
     "myusernamehere", ENV['SIDEKIQ_PASSWORD']
   ]
end
mount Sidekiq::Web => '/sidekiq'
```

**Comment by jhollinger** [comment]
We're using latest sinatra (1.4.6) and rack-protection (1.5.3) and this was still an issue. We had to manually create a session cookie and initialize a CSRF token ourselves with some custom middleware. I don't understand why Sidekiq::Web/Sinatra can't do this for us - or least have an option to do it. Because from what I can tell anyone who doesn't otherwise have some other Rack app creating a standard Rails session cookie will need to do this. Seems unnecessary to push that off to the developer when in those cases, Sidekiq::Web is the *only* thing that needs that cookie.

Anyway, here's what our `config.ru` looks like now, and a link to our CSRF-inserting middleware.

    ...
    map '/sidekiq' do
      use Rack::Session::Cookie, key: 'sidekiq.session', secret: ENV['SESSION_SECRET']
      use Sidekiq::CsrfTokenFix
      run Sidekiq::Web
    end
    ...

https://gist.github.com/jhollinger/c41011417f729ffe5c5c

**Comment by mperham** [comment]
@jhollinger Try this instead of your code:

```ruby
map '/sidekiq' do
  require 'rack/protection'
  use Rack::Protection::AuthenticityToken
  run Sidekiq::Web
```

Your stack must not be running rack-protection since it initializes the csrf token here:

https://github.com/sinatra/rack-protection/blob/master/lib/rack/protection/authenticity_token.rb#L24

**Comment by mperham** [comment]
I also want to point out to everyone that Sidekiq has a very simple Rack endpoint running Sidekiq::Web which does work in my testing.  You can use it to compare with your own environment to debug.

```bash
# git clone mperham/sidekiq
cd sidekiq/myapp
bundle
bundle exec rackup simple.ru
open http://localhost:9292
# go to Queues page, click Delete on the default queue
```

**Comment by jhollinger** [comment]
That fails with `Puma caught this error: you need to set up a session middleware *before* Rack::Protection::AuthenticityToken (RuntimeError)`. But it works when I add back a session cookie:

    map '/sidekiq' do
      use Rack::Session::Cookie, key: 'sidekiq.session', secret: ENV['SESSION_SECRET']
      use Rack::Protection::AuthenticityToken
      run Sidekiq::Web
    end

So that's am improvement - don't need to write custom middleware. But to me it looks superficially similar to https://github.com/mperham/sidekiq/blob/v3.5.3/lib/sidekiq/web.rb#L14. But maybe it's not quite the same?

**Comment by mperham** [comment]
@jhollinger Good info.  If I say `enable :sessions`, Sinatra is suppose to supply a working session to the app.  If your environment is not doing that, something is wrong.  You might want to step into that line executing in a debugger.  I'd bet you have some other gem hacking up Sinatra/Rack.

**Comment by matisoffn** [comment]
I'm having the same problem, but my app is running on Heroku.  I haven't changed my Heroku setup.  Has anyone had the same instance occur?

**Comment by inestyne** [comment]
Same issue, but mine is after upgrading rails to 2.4.5.

routes.rb:

require 'sidekiq/web'

WebSiteCom::Application.routes.draw do
  root :to => "home#index"
  mount Sidekiq::Web => '/sidekiq'

**Comment by tanelsuurhans** [comment]
Dug into this issue a little more by trying to trace down where things go wrong - because I can clearly see the session being present and a CSRF value being store, but for some reason the values don't match (the one from the session vs the one submitted with a form). 

The way Sinatra handles sessions when you just say `enable :sessions` is to use Rack::Session::Cookie and generate a `:secret` as an option for it using SecureRandom.hex. This all works fine and dandy, but what I'm seeing is this piece of code being ran twice with different process ID-s. Because the value then changes across different requests, the old value already rendered with buttons and forms is essentially invalidated and this is what causes the error when you submit any of those. If the cookie secret value is hardcoded inside `Sidekiq::Web`, everything works as expected - but this obviously is not feasible.

The setup I'm running this with locally in development mode is a Rails application with POW, behind a simple Nginx proxy. The main application is using Redis for sessions (but I don't think this makes any difference). I can reliably replicate this with the given setup. We also have the same issue in production, where we run multiple nodes running 10 Unicorn workers per application. The same issue happens there as well and logically it would be caused by different nodes having a different `:secret` key generated on the application startup. So now when node #1 submits the form to node #2, the values do not match and CSFR protection kicks in. This seems to be validated by not being able to reproduce the same behavior on our staging environment, which consists of one server with 10 workers.

As far as nice and clean fixes go, I'm still somewhat stuck. Initially I figured adding this to my initializer would help:
`Sidekiq::Web.use Rack::Session::Cookie, :secret => Settings.sidekiq_secret `
The problem is that this has no effect whatsoever unless it's done in config.ru and Sidekiq::Web mounted via that method - which in my case undesirable. 

@mperham is correct in the sense that when enabling sessions for Sinatra, its supposed to provide a working session, and it does, but it doesn't really work in a multi-process environment apparently. What we need is a nice way to optionally configure the sessions secret via the "owner" Rails application. This would allow to avoid having to mess around with separate rackup files etc. Is there a way to expose a configuration setting like that? Perhaps even defaulting to the Rails session secret if its present?

**Comment by mperham** [comment]
@tanelsuurhans FYI this is the latest find: https://github.com/mperham/sidekiq/issues/2730

**Comment by tanelsuurhans** [comment]
Ha @mperham, while it directly does not solve the issue - as its not wildcard domain session related - it does contain the solution that I was looking for. I can set the missing session value for all nodes like this:
`Sidekiq::Web.set :session_secret, Settings.sidekiq_secret`

So instead of trying to make it use Rack::Session::Cookie once more (which had no effect), I can set the cookie secret value instead. At first glance it seems to be working locally, have to test it out in production too to be sure. Thanks.

**Comment by aquach** [comment]
I just dug into this issue to figure out why this was happening. It turns out that I had two Session::Cookie middlewares in the stack, which were fighting each other to set cookies (you can observe this with Chrome inspector — the HTTP responses will sometimes have two Set-Cookie headers). One was coming from Sinatra when Sidekiq::Web calls `enable :sessions`, and one comes from my `config.ru`, which was calling `use Rack::Session::Cookie, secret: <stuff>` per https://github.com/mperham/sidekiq/wiki/Monitoring#rack-session-and-protection-against-web-attacks, #2487, and #2630.

From this, it seems like the `use Session::Cookie` is actually extraneous, because Sinatra is already configuring the cookie just fine. When `enable :sessions` is called, Sinatra adds the middleware and uses a randomly generated secret, which is fine for single-process Sidekiq web instances because the cookie is only used to prevent CSRF attacks rather than to allow people to login (otherwise, every server restart would log everybody out).

This seems contrary to all the advice given in other issues with this (such as "Sinatra must have a working session store. I don't know how to enable that for your specific environment" from #2487) given that leaving Sinatra alone to do its thing seems to get the job done. Perhaps the changelog and documentation should be amended to indicate that explicitly adding the Cookie middleware actually breaks the system.

Incidentally, it looks like only the `authenticity_token` component of rack-protection is being used, so it's unclear to me why people managed to fix their "Forbidden" issues by adding nginx directives because only modules like Rack::Protection::HttpOrigin care about those (#2560).

Another note: #2587 led me to the `myapp` folder in Sidekiq, which, contrary to the general documentation to `use Rack::Session::Cookie`, doesn't actually do so. That led me to realize that adding the middleware manually was actually unnecessary, because I was confused why `myapp` was not exhibiting the same "Forbidden" errors that my configuration was.

In short, if you're seeing "Forbidden" issues and you have `use Rack::Session::Cookie` in your configuration, remove it. You also don't need to rotate your middleware or other shenanigans (#2459). If you want to configure your session secret yourself, use `Sidekiq::Web.set :session_secret, "SECRET"` as demonstrated above.

**Comment by mperham** [comment]
@aquach Great summary.  My takeaway: if you have problems, remove any Sinatra customization in your code except for `Sidekiq::Web.set :session_secret, "SECRET"`  Thank you for your hard work.

**Comment by benhoxie** [comment]
TY @aquach , your link to #2560 fixed this for me very quickly on Rails 4.2/sidekiq 4/ puma/nginx.

**Comment by glaucocustodio** [comment]
When using `MyRailsApp::Application.config.session_store :redis_store` I had to set `Sidekiq::Web.use ActionDispatch::Session::RedisStore` in `routes.rb`.

**Comment by gizipp** [comment]
```
map '/sidekiq' do
  use Rack::Session::Cookie, key: 'sidekiq.session', secret: ENV['SESSION_SECRET']
  use Rack::Protection::AuthenticityToken
  run Sidekiq::Web
end
```

Thanks a bunch @jhollinger

**Comment by tanelsuurhans** [comment]
I just ended up doing this - seems way easier:
`Sidekiq::Web.set :session_secret, <your secret>`

**Comment by 31piy** [comment]
I had a similar issue in my app. I could not perform non-GET operations on sidekiq after using nginx as a reverse proxy. The app was deployed on three different hosts, namely UI (NodeJS), API (Rails) and Load Balancer (Nginx).

I made the following changes in order to make that work:

In `config/routes.rb`, I added the following line after `require 'sidekiq/web'`
```ruby
Sidekiq::Web.set :session_secret, Rails.application.secrets[:secret_key_base]
```
In `nginx.conf`, I added two more headers
```
proxy_set_header        X-Forwarded-Host 'my-host.com';
proxy_set_header        X-Forwarded-Proto $scheme;
```
The latter was required because Rack Protection uses the X-Forwarded-Host header to protect against CSRF attacks. Refer to line 16 of `rack/protection/remote_referrer.rb` in version 1.5.3.

**Comment by Asoul** [comment]
Thanks @31piy,
```
In nginx.conf, I added two more headers

proxy_set_header        X-Forwarded-Host 'my-host.com';
proxy_set_header        X-Forwarded-Proto $scheme;
```

This work for me !

**Comment by bbozo** [comment]
Did anyone maybe figure out an apache equivalent of this config? Same issue here, rails + active admin + sidekiq web ui, `Sidekiq::Web.set :session_secret, Rails.application.secrets[:secret_key_base]` didn't fix the issue, so now I'm suspecting the reverse proxy Apache

```
proxy_set_header        X-Forwarded-Host 'my-host.com';
proxy_set_header        X-Forwarded-Proto $scheme;
```

**Comment by bradurani** [comment]
Has anyone found a fix for Heroku?
I'm getting this on Rails 4.2.7.1
Puma 3.6.0 
ruby 2.3.1-p112
Sidekiq 4.2.2
```
#session_store.rb
Rails.application.config.session_store :cookie_store
```
```
# routes.rb
require 'sidekiq/web'
mount Sidekiq::Web => '/sidekiq'
```

**Comment by bradurani** [comment]
@matisoffn Did you ever find a fix for Heroku?

**Comment by 31piy** [comment]
@bradurani What I mentioned [here](https://github.com/mperham/sidekiq/issues/1289#issuecomment-232330804) also runs on heroku.

**Comment by bradurani** [comment]
Turns out this was only an issue on Chrome. I switched to Firefox and it's not an issue. Anyone know why?

**Comment by avongluck-r1soft** [comment]
This is what finally solved it for me:
```
proxy_set_header        X-Forwarded-Host $host;
proxy_set_header        X-Forwarded-Proto $scheme;
```

and just a simple mount:
```
  constraints admin_constraint do
    mount Sidekiq::Web => '/sidekiq'
  end
```

**Comment by bradurani** [comment]
@31piy you run nginx on Heroku?

**Comment by 31piy** [comment]
@bradurani Yes. We have a micro rails app in combination with three heroku buildpacks:

1. https://github.com/dpiddy/heroku-buildpack-runit.git
2. heroku/ruby
3. https://github.com/ryandotsmith/nginx-buildpack.git

Please refer to the above git repos for more info.

**Comment by nitzanav** [comment]
This worked for me:
```
Sidekiq::Web.set :session_secret, ENV['SESSION_SECRET']
run Sidekiq::Web
```

But I wonder about this code:
```
    def build
      middlewares = self.middlewares
      klass = self.class

      unless using?(::Rack::Protection) || ENV['RACK_ENV'] == 'test'
        middlewares.unshift [[::Rack::Protection, { use: :authenticity_token }], nil]
      end

      unless using? ::Rack::Session::Cookie
        unless secret = Web.session_secret
          require 'securerandom'
          secret = SecureRandom.hex(64)
        end

        middlewares.unshift [[::Rack::Session::Cookie, { secret: secret }], nil]
      end
```

Shouldn't `secret = SecureRandom.hex(64)` when using CSRF with multiple process worker (e.g. unicorn or puma)? I mean when on worker process generates the view and another worker get the post request while both has different random secret.

**Comment by pioz** [comment]
Someone know how to add this settings
```
proxy_set_header X-Forwarded-Host $host;
proxy_set_header X-Forwarded-Proto $scheme;
```
if you are using Passenger in standalone mode?

**Comment by pioz** [comment]
If you are using Apache add
```
RequestHeader set X-Forwarded-Proto "https"
RequestHeader set X-Forwarded-Host "example.com"
```
in your vhost conf file. Worked for me!

**Comment by pnomolos** [comment]
To those fighting with this, I've added some notes on the wiki to hopefully aid debugging: https://github.com/mperham/sidekiq/wiki/Monitoring#forbidden

It'd be nice to have a larger list of common pitfalls and how to solve them.

**Comment by heaven** [comment]
```
Started POST "/sidekiq/queues/default" for 127.0.0.1 at 2017-09-14 20:26:31 +0300
  Employee Load (0.3ms)  SELECT  `users`.* FROM `users`...
  UserRole Load (0.4ms)  SELECT `user_roles`.* FROM `user_roles` ...
W, [2017-09-14T20:26:31.651999 #35081]  WARN -- : attack prevented by Rack::Protection::AuthenticityToken
   (0.2ms)  BEGIN
  SQL (0.4ms)  UPDATE `sessions` SET `data` = 'BAh...AEY=\n' WHERE `sessions`.`id` = 560409
   (0.4ms)  COMMIT
```
Previously we had `Sidekiq::Web.set :sessions, { domain: ".#{Settings.host}" }` there which worked with Rails 4, but stopped in Rails 5. Only this `Sidekiq::Web.set :sessions, false` helped.

**Comment by Mexxerio** [comment]
I'll just leave this here in case someone stumbles on here while trying to find a solution and maybe it can save you hours of trial and error with different nginx headers:

In addition to setting these nginx headers:
```
proxy_set_header Host $host;
proxy_set_header X-Forwarded-Proto $scheme;
proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
```

You also have to add the following to either `routes.rb` or `config.ru` if you're getting a HttpOrigin Error from Rack Protection

```ruby
Sidekiq::Web.class_eval do
  use Rack::Protection, origin_whitelist: ['https://your.domain'] # resolve Rack Protection HttpOrigin
end
```

That did the trick for me.

**Comment by unavailabl3** [comment]
This link helped me
[https://github.com/mperham/sidekiq/wiki/Monitoring#proxying-the-web-ui](https://github.com/mperham/sidekiq/wiki/Monitoring#proxying-the-web-ui)
